apiserver/facades/client/modelupgrader/controllerupgrader.go

```diff
@@ -0,0 +1,278 @@
+	// upgraded to the specified version. It performs validation checks to ensure that
+	// the target version is valid and that the upgrade can proceed.
+	CanUpgradeControllerToVersion(ctx context.Context, desiredVersion semversion.Number) error
+
+	// CanUpgradeControllerWithStream determines whether the controller can be upgraded
+	// to the latest available patch version within the specified agent stream. It returns
+	// the desired version that the controller can upgrade to if all validation checks pass.
+	CanUpgradeControllerWithStream(ctx context.Context, stream modelagent.AgentStream) (semversion.Number, error)
+
+	// CanUpgradeControllerToVersionWithStream determines whether the controller can be
+	// safely upgraded to the specified version within the given agent stream.
+	CanUpgradeControllerToVersionWithStream(ctx context.Context, desiredVersion semversion.Number, stream modelagent.AgentStream) error
+}
+
+// ControllerUpgraderAPI upgrades a controller and a model hosting the controller.
+type ControllerUpgraderAPI struct {
+	authorizer facade.Authorizer
+	check      common.BlockCheckerInterface
+
+	upgraderService ControllerUpgraderService
+
+	controllerTag names.Tag
+	modelTag      names.Tag
+}
+
+// NewControllerUpgraderAPI instantiates a new [ControllerUpgraderAPI].
+func NewControllerUpgraderAPI(
+	controllerTag names.Tag,
+	modelTag names.Tag,
+	authorizer facade.Authorizer,
+	check common.BlockCheckerInterface,
+	upgraderService ControllerUpgraderService,
+) *ControllerUpgraderAPI {
+	return &ControllerUpgraderAPI{
+		controllerTag:   controllerTag,
+		modelTag:        modelTag,
+		authorizer:      authorizer,
+		check:           check,
+		upgraderService: upgraderService,
+	}
+}
+
+// AbortModelUpgrade returns not supported, as it's not possible to move
+// back to a prior version.
+func (c *ControllerUpgraderAPI) AbortModelUpgrade(_ context.Context, _ params.ModelParam) error {
+	return errors.New("aborting model upgrades is not supported").Add(coreerrors.NotSupported)
+}
+
+// canUpgrade has the responsibility to determine whether there is sufficient permission
+// to perform an upgrade.
+func (c *ControllerUpgraderAPI) canUpgrade(ctx context.Context, model names.ModelTag) (bool, error) {
+	if model.Id() != c.modelTag.Id() {
+		return false, nil
+	}
+	err := c.authorizer.HasPermission(
+		ctx,
+		permission.SuperuserAccess,
+		c.controllerTag,
+	)
+	if err != nil && !errors.Is(err, authentication.ErrorEntityMissingPermission) {
+		return false, errors.Capture(err)
+	}
+	if err == nil {
+		return true, nil
+	}
+
+	err = c.authorizer.HasPermission(
+		ctx,
+		permission.WriteAccess,
+		model,
+	)
+	if err != nil {
+		if errors.Is(err, authentication.ErrorEntityMissingPermission) {
+			return false, nil
+		}
+		return false, errors.Capture(err)
+	}
+	return true, nil
+}
+
+func (c *ControllerUpgraderAPI) mapResponse(err error, targetVersion semversion.Number, arg params.UpgradeModelParams) (params.UpgradeModelResult, error) {
```

Comment:
After reading the code I would be changing this func name to `handleUpgradeErr` or something similar.